src/interaction/sessionrecording.cpp

```diff
@@ -0,0 +1,1207 @@
+            keyframeLine << std::fixed << std::setprecision(3) << global::timeManager.time().j2000Seconds();
+            keyframeLine << " ";
+            //Add camera position
+            keyframeLine << std::fixed << std::setprecision(7) << kf._position.x << " "
+                << std::fixed << std::setprecision(7) << kf._position.y << " "
+                << std::fixed << std::setprecision(7) << kf._position.z << " ";
+            //Add camera rotation
+            keyframeLine << std::fixed << std::setprecision(7) << kf._rotation.x << " "
+                << std::fixed << std::setprecision(7) << kf._rotation.y << " "
+                << std::fixed << std::setprecision(7) << kf._rotation.z << " "
+                << std::fixed << std::setprecision(7) << kf._rotation.w << " ";
+            keyframeLine << std::fixed << std::setprecision(7) << kf._scale << " ";
+            if (kf._followNodeRotation)
+                keyframeLine << "F ";
+            else
+                keyframeLine << "- ";
+            keyframeLine << kf._focusNode;
+
+            saveKeyframeToFile(keyframeLine.str());
+        }
+    }
+}
+
+void SessionRecording::saveTimeKeyframe() {
+    if (_state != sessionState::recording)
+        return;
+
+    //Create a time keyframe, then call to populate it with current time props
+    datamessagestructures::TimeKeyframe kf;
+    _externInteract.generateTimeKeyframe(kf);
+
+    if (isDataModeBinary()) {
+        _bufferIndex = 0;
+        _keyframeBuffer[_bufferIndex++] = 't';
+        writeToFileBuffer(kf._timestamp);
+        writeToFileBuffer(kf._timestamp - _timestampRecordStarted);
+        writeToFileBuffer(kf._time);
+        writeToFileBuffer(kf._dt);
+        writeToFileBuffer(kf._paused);
+        writeToFileBuffer(kf._requiresTimeJump);
+
+        saveKeyframeToFileBinary(_keyframeBuffer, _bufferIndex);
+    } else {
+        std::stringstream keyframeLine = std::stringstream();
+        //Add simulation timestamp, timestamp relative, simulation time to recording start
+        keyframeLine << "time ";
+        keyframeLine << kf._timestamp << " ";
+        keyframeLine << (kf._timestamp - _timestampRecordStarted) << " ";
+
+        //keyframeLine << std::fixed << std::setprecision(3) << global::timeManager.time().j2000Seconds();
+        keyframeLine << std::fixed << std::setprecision(3) << kf._time;
+
+        keyframeLine << " " << kf._dt;
+        if (kf._paused)
+            keyframeLine << " P";
+        else
+            keyframeLine << " R";
+
+        if (kf._requiresTimeJump)
+            keyframeLine << " J";
+        else
+            keyframeLine << " -";
+
+        saveKeyframeToFile(keyframeLine.str());
+    }
+}
+
+void SessionRecording::saveScriptKeyframe(std::string scriptToSave) {
+    if (_state != sessionState::recording)
+        return;
+
+    datamessagestructures::ScriptMessage sm;
+    _externInteract.generateScriptMessage(sm, scriptToSave);
```

Comment:
Can use `std::move` here to avoid the copy

Reply:
The generateScriptMessage function was changed as a result of a previous change in this review. sm no longer passed-in, and the scriptToSave is std::moved inside the function